### ISSUE TITLE
bug(cuda): graph decode replay crashes — cudaMalloc inside stream capture (blocks #932) on A100 sm_80

### ISSUE THREAD

**Issue by anthony-chaudhary** [anthony-chaudhary/fak#969]
## Summary

On a real A100 (sm_80) GPU server, the CUDA-graph decode path crashes during stream
capture: a device allocation (`cudaMalloc` / `cudaMallocManaged`) happens **inside** an
active CUDA-graph stream capture, which CUDA forbids. This makes the `-tags cuda` graph
witnesses red and panics the throughput benchmarks, so the graph-decode replay cannot be
measured. This directly **blocks #932** (witness `--cuda-graph` decode replay at 27B).

## Evidence (measured 2026-06-27, A100-SXM4-80GB, sm_80, CUDA 12.8, fak HEAD 265ce03)

`go test -tags cuda` run per-test in isolation:

```
FAIL TestCUDAGraphDecodeParity ::
  fak-cuda: cudaMalloc(256 bytes) failed: operation not permitted when stream is capturing
  fak-cuda: cudaMallocManaged(256 bytes) failed: operation not permitted when stream is capturing
```

The throughput benchmark panics for the same reason and fails the whole `internal/compute`
bench package:

```
panic ... internal/compute/cuda_graph_test.go:158
  BenchmarkCUDADecodeNoCapture / BenchmarkCUDADecodeCapture
exit status 2
FAIL github.com/anthony-chaudhary/fak/internal/compute
```

Every other device-parity witness on the same box PASSES (AWQ, Q8, Q4_K, F16, flash-attn,
full forward, model forward — cosine 1.0 / 0.9999), so this is isolated to the graph path.

## Root cause

A 256-byte device buffer is being allocated lazily on the decode hot path while the CUDA
stream is mid-capture. CUDA disallows `cudaMalloc`/`cudaMallocManaged` during capture
(`cudaErrorStreamCaptureUnsupported` / "operation not permitted when stream is capturing").

## Fix direction

Pre-allocate **all** device scratch the graphed decode step can touch (argmax scratch, the
256-byte temp, any managed fallback) **before** `cudaStreamBeginCapture`, and assert no
allocator call happens between begin/end capture. Then re-run `TestCUDAGraphDecodeParity`
and `BenchmarkCUDADecode{Capture,NoCapture}` to get the graph-replay speedup #932 wants.

## Acceptance

- `go test -tags cuda -run TestCUDAGraphDecodeParity ./internal/compute` passes on sm_80.
- `BenchmarkCUDADecodeCapture` vs `NoCapture` both run (no panic) and yield a recorded
  ns/op delta = the graph-replay win (feeds #932 / #483).

/cc #932 #483

**Comment by anthony-chaudhary** [comment]
**Code pointers for the fix** (from a read of `internal/compute/cuda.go`):

The path already has warm-up machinery that is *supposed* to guarantee no allocation during capture:
- `cuda.go:367` — "Preconditions the HAL guarantees: pool warm (no cudaMalloc during capture) + fixed-capacity KV".
- `AppendKV` pre-allocates a fixed `cudaKVMaxPos` capacity so there's no mid-token `cudaMalloc` (`cuda.go:1154-1167`).
- The benchmarks `Recycle()` + a warm `stepDev()` before `ResetTimer()` to populate the buffer pool.

Despite that, a **256-byte** device buffer still gets allocated *inside* capture on sm_80, and it goes through `dalloc` (`cuda.go:453`: `fcuda_malloc`, then the `fcuda_malloc_managed` fallback at `:455`) — which is why the failure prints **both** `cudaMalloc` and `cudaMallocManaged` "operation not permitted when stream is capturing": dalloc tries pooled then managed, both illegal mid-capture.

So the warm-up is missing one small (256B ≈ 64 f32) scratch/argmax buffer that the captured decode step touches lazily. Fix = make the warm `stepDev()` (or an explicit pre-capture reservation) allocate that buffer too, and add an assert that `dalloc` is never reached while a capture is open. Repro per-test:
`go test -tags cuda -run TestCUDAGraphDecodeParity ./internal/compute` on an A100 sm_80.

**Comment by anthony-chaudhary** [comment]
Threaded under the GPU decode-parity playbook epic **#977**.

In the #67 method, the "one submit per forward" lever on CUDA **is** the CUDA graph — so this `cudaMalloc-inside-stream-capture` crash is the gate that has to clear before the playbook can even start on this backend: until graph replay is captured cleanly, there is no single-command-buffer forward to profile. Fix the lazy 256-byte device alloc on the decode hot path (pre-allocate the scratch outside capture), then #932 can witness the `--cuda-graph` decode replay at 27B and the "find the serial step hiding behind the parallel one" step (#977's real work) can begin.

**Comment by anthony-chaudhary** [comment]
Pushed a layered fix in `35f0d296` (`fix(compute): prevent CUDA-graph capture crashes by preallocating cuBLAS workspace + decode scratch`).

**Root cause this fix targets:** the prior warm-fix (#932, `d7d61725`) warmed *fak's own* transient pool before capture, but **cuBLAS allocates its internal workspace via its own `cudaMalloc` on the first GEMM problem it sees** — an allocation pool-warming can't reach. The captured F32 decode GEMMs (`TestCUDAGraphDecodeParity` / `decodeChain`) trip that lazy alloc *inside* capture. The 256-byte size + the `cudaMallocManaged` twin in the report are the cuBLAS-workspace signature, not fak's ~600 KB vocab logits buffer.

**Fix:**
- `fcuda_init` now allocates an explicit 32 MiB cuBLAS workspace and hands it to `cublasSetWorkspace`, so cuBLAS uses that buffer exclusively and never allocates during capture (the documented CUDA remedy; 32 MiB is NVIDIA's Hopper size, a safe superset on Ampere/Ada). **This is the root-cause fix for the F32 graph test.**
- The Q8 activation-quant scratch (`fcuda_q8_matmul_f32`) and the argmax index buffer are now persistent grow-once statics (mirroring `g_attn_scratch`) instead of per-call `fcuda_malloc`/`fcuda_free`, so fak's own pooled allocator can't miss → `cudaMalloc` on a fresh size class mid-capture for the real Q8 decode path.

`internal/compute` builds green (non-cuda), gofmt clean. **Not yet verified on hardware** — the `-tags cuda` build + re-running `TestCUDAGraphDecodeParity` and `BenchmarkCUDADecode{Capture,NoCapture}` on the A100 (sm_80) is the next step to confirm the crash is gone and record the graph-replay tok/s delta for #932.

**Comment by anthony-chaudhary** [comment]
Resolved by `35f0d29` (resolving commit). Closed by the DOS dispatch loop's close-resolved arm, witnessed via `dos commit-audit` (verdict OK / diff-witnessed). Reopen if this does not fully resolve it.

**Comment by anthony-chaudhary** [comment]
Residual fix committed in e4fab9cf (fak compute): added a g_capture_open flag (set in fcuda_graph_begin, cleared in end_launch/abort) asserted in fcuda_malloc — a device alloc on the captured decode path now fails LOUD naming the un-pre-warmed size instead of crashing with the opaque 'operation not permitted when stream is capturing'. Also fixed a stale comment referencing a non-existent fcuda_graph_warm(). The byte-level pre-warm (cuBLAS workspace prealloc + persistent g_q8_*/g_argmax scratch + the uncaptured halLogitsWarm step) is already on trunk. NOT YET re-witnessed: needs an A100 sm_80 run of `go test -tags cuda -run TestCUDAGraphDecodeParity ./internal/compute/` to confirm capture/replay completes without the crash.

**Comment by anthony-chaudhary** [comment]
**FIXED — verified on sm_80 hardware.** `TestCUDAGraphDecodeParity` now PASSES on the sm_80 datacenter-GPU bench node:

```
--- PASS: TestCUDAGraphDecodeParity (0.10s)
#483 graph witness: 4 prompt + 8 greedy steps, graph-replayed==eager argmax-exact, logit cosine >= 0.999; device=cuda tier=sm_80 class=approx
ok  github.com/anthony-chaudhary/fak/internal/compute  0.778s
```

The crash was a **pool-depth** problem, not the allocators I first suspected. The on-hardware witness ruled those out: after the cuBLAS-workspace + persistent-Q8/argmax fixes (`35f0d296`), the crash *moved* to `RMSNorm → devTr → dallocClass(256B)` — the C guard pinpointed it as `fcuda_malloc(256) on a pool miss while a CUDA graph capture is OPEN`.

Root cause: the warm forward pools ONE buffer per transient size, but a single captured `decodeChain` holds SEVERAL same-size transients live at once (the per-layer RMSNorm outputs are all 256B and none recycle mid-forward), so the pool drains mid-forward and the next same-size `devTr` misses → `cudaMalloc`, illegal during capture.

**Fix (`0168bb08` + `f7adbacc`):** `fcuda_graph_prewarm(extra)` tops every pool bucket up to `extra` spares, called from `GraphBegin` right before `fcuda_graph_begin` (outside capture). `graphPrewarmDepth=128` covers the peak same-size concurrency for any realistic layer count at negligible VRAM. (`f7adbacc` added the missing `cuda_backend.h` prototype — the first build was red with `could not determine what C.fcuda_graph_prewarm refers to`.)

This unblocks #932 (graph-replay decode). Next: `BenchmarkCUDADecode{Capture,NoCapture}` for the graph-replay tok/s delta. Closing — the parity witness is green.
